cmd/vulcan-subdomain-takeover/main.go

```diff
@@ -0,0 +1,381 @@
+		resp, err := paginator.NextPage(context.Background())
+		if err != nil {
+			return nil, fmt.Errorf("list hosted zones: %w", err)
+		}
+		hostedZones = append(hostedZones, resp.HostedZones...)
+	}
+	return hostedZones, nil
+}
+
+// getRoute53ZoneRecords retrieves all the Zone Records for a ZoneId.
+func (s Scanner) getRoute53ZoneRecords(zoneId *string) ([]types.ResourceRecordSet, error) {
+	listParams := &route53.ListResourceRecordSetsInput{
+		HostedZoneId: zoneId,
+	}
+	paginator := route53.NewListResourceRecordSetsPaginator(s.route53Client, listParams)
+	var zoneRecords []types.ResourceRecordSet
+	for paginator.HasMorePages() {
+		resp, err := paginator.NextPage(context.Background())
+		if err != nil {
+			return nil, fmt.Errorf("list resource record sets: %w", err)
+		}
+		zoneRecords = append(zoneRecords, resp.ResourceRecordSets...)
+	}
+	return zoneRecords, nil
+}
+
+// getIPs retrieve the public IPs.
+func (s Scanner) getIPs() ([]string, error) {
+	var elasticIPs = make(map[string]string)
+	// Get the public IPs from the addresses.
+	paginatorAddresses := ec2.NewDescribeAddressesAttributePaginator(s.ec2Client, nil)
+	for paginatorAddresses.HasMorePages() {
+		resp, err := paginatorAddresses.NextPage(context.Background())
+		if err != nil {
+			return nil, fmt.Errorf("describe addresses: %w", err)
+		}
+		for _, address := range resp.Addresses {
+			elasticIPs[*address.PublicIp] = *address.PublicIp
+		}
+	}
+
+	// Get the public IPs from the Network Interfaces.
+	paginatorNetworkInterfaces := ec2.NewDescribeNetworkInterfacesPaginator(s.ec2Client, nil)
+	for paginatorNetworkInterfaces.HasMorePages() {
+		resp, err := paginatorNetworkInterfaces.NextPage(context.Background())
+		if err != nil {
+			return nil, fmt.Errorf("describe network interfaces: %w", err)
+		}
+		for _, networkInterface := range resp.NetworkInterfaces {
+			if networkInterface.Association != nil {
+				elasticIPs[*networkInterface.Association.PublicIp] = *networkInterface.Association.PublicIp
+			}
+		}
+	}
+	return slices.Collect(maps.Keys(elasticIPs)), nil
+}
+
+// calculateTakeovers crosses the dnsRecords with the elasticIPs and determine which of the
+// dnsRecords are dangling.
+func (s Scanner) calculateTakeovers(dnsRecords []dnsRecord, elasticIPs []string) (map[string]string, error) {
+	// find all DNS records that point to EC2 IP addresses.
+	dnsEC2IPs := make(map[string][]string)
+	aip, err := s.ipRangesClient.GetPrefixes()
+	if err != nil {
+		return nil, fmt.Errorf("get ip ranges: %w", err)
+	}
+
+	for _, dnsr := range dnsRecords {
+		for _, record := range dnsr.records {
+			awsMetadata, err := aip.GetPrefixByIP(record)
+			if err != nil {
+				if errors.Is(err, ErrPrefixNotFound) {
+					continue
+				}
+				return nil, fmt.Errorf("getting AWS metadata for %s: %w", record, err)
+			}
+			for _, service := range strings.Split(awsMetadata.Service, ",") {
+				if service == "EC2" {
+					dnsEC2IPs[dnsr.name] = dnsr.records
+				}
+			}
+		}
+	}
```

Comment:
Why not just use a regex.MustCompile(`\bEC2\b`) for testing that without a loop?